File: controllers/src/main/java/com/example/controllers/characterdetail/CharacterDetailViewModelDelegate.java
package com.example.controllers.characterdetail;

import android.util.Log;

import com.example.controllers.commons.ProcessedMarvelCharacter;
import com.example.controllers.commons.ProcessedMarvelComic;
import com.example.mviframework.BaseMviDelegate;
import com.example.mviframework.Change;
import com.example.mviframework.Reducer;
import com.example.mviframework.Runner;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;

public class CharacterDetailViewModelDelegate extends BaseMviDelegate<State, CharacterDetailViewModelDelegate.InnerState, Effect> {
    CharacterDetailNetworkInterface networkInterface;
    ProcessedMarvelCharacter character;
    public CharacterDetailViewModelDelegate(CharacterDetailNetworkInterface networkInterface, ProcessedMarvelCharacter character) {
        this.networkInterface = networkInterface;
        this.character = character;
        loadComics(character.id);
        /*enqueue(new Reducer<InnerState, Effect>() {
            @Override
            public Change<InnerState, Effect> reduce(InnerState innerState) {
                InnerState newInnerState = innerState.copy();
                newInnerState.character loadComics= character;
                newInnerState.error = false;
                newInnerState.loading = false;
                newInnerState.isImageLoaded = false;

                return asChange(newInnerState);
            }
        });*/
    }



    Runner callbackRunner = dispatch(new Reducer<InnerState, Effect>() {
        @Override
        public Change<InnerState, Effect> reduce(InnerState innerState) {
            InnerState newInnerState = innerState.copy();
            newInnerState.loading = false;
            newInnerState.error = false;
            return withEffects(newInnerState, new Effect.ImageLoaded());
            //return asChange(newInnerState);
        }
    });

   /* ImageLoadedCallbackListener callbackListener = new ImageLoadedCallbackListener() {
        @Override
        public void callback() {
            enqueue(new Reducer<InnerState, Effect>() {
                @Override
                public Change<InnerState, Effect> reduce(InnerState innerState) {
                    InnerState newInnerState = innerState.copy();
                    newInnerState.isImageLoaded = true;
                    return asChange(newInnerState);
                }
            });
        }
    };*/

    @Override
    public Change<InnerState, Effect> getInitialChange() {
        return asChange(new InnerState(character, null,true, false));
    }

    @Override
    public State mapState(InnerState innerState) {
        Log.d("Vartika3", "comics in mapstate"+innerState.comics);
        return new State(innerState.character, innerState.comics, innerState.loading, innerState.error, callbackRunner);
    }

    static class InnerState {
        boolean loading;
        boolean error;
        ProcessedMarvelCharacter character;
        List<ProcessedMarvelComic> comics;
        InnerState(ProcessedMarvelCharacter character, List<ProcessedMarvelComic> comics, boolean loading, boolean error) {
            this.character = character;
            this.loading = loading;
            this.error = error;
            this.comics = comics;
        }
        InnerState copy() {
            return new InnerState(character, comics, loading, error);
        }
    }

    private void loadComics(int characterId) {
        if (networkInterface == null) {
            return;
        }
        Observable<List<ProcessedMarvelComic>> observable = networkInterface
                            .loadCharacterComics(characterId)
                            .onErrorReturn(throwable -> {
                                return new ArrayList<>();
                            });



        Observable<Reducer<InnerState, Effect>> reducerObservable = observable
                .map(comics -> {
                    Log.d("Vartika3", "Reducer1: "+comics);
                    return new Reducer<InnerState, Effect>() {

                        @Override
                        public Change<InnerState, Effect> reduce(InnerState innerState) {
                            Log.d("Vartika3", "Reducer2: "+comics);
                            InnerState newInnerState = innerState.copy();
                            newInnerState.comics = comics;
                            return asChange(newInnerState);
                        }
                    };
                });
        enqueue(reducerObservable);
    }

}

File: controllers/src/main/java/com/example/controllers/characterdetail/CharacterDetailViewModelDelegate.kt
package com.example.controllers.characterdetail

import android.util.Log
import android.view.View
import com.example.controllers.commons.*
import com.example.mviframework.BaseMviDelegate
import com.example.mviframework.Change
import com.example.mviframework.Reducer
import java.util.*

class CharacterDetailViewModelDelegate(val networkInterface: CharacterDetailNetworkInterface?, val character: ProcessedMarvelCharacter) : BaseMviDelegate<State, CharacterDetailViewModelDelegate.InnerState, Effect>() {
    var cardClickListener: CardClickListener<*> = object : CardClickListener<ProcessedMarvelItemBase> {
        override fun invoke(view: View, item: ProcessedMarvelItemBase) {
            enqueue { innerState -> withEffects(innerState, Effect.ClickCardEffect(view, item)) }
        }
    }
    var callbackRunner = dispatch { innerState ->
        val newInnerState = innerState.copy()
        newInnerState.loading = false
        newInnerState.error = false
        withEffects(newInnerState, Effect.ImageLoaded<Any?>())
    }

    fun onUrlLinkClicked(url: String) {
        enqueue {
            withEffects(it, Effect.OpenUrl(url))
        }
    }

    var shareListener: () -> Unit = {
        enqueue { innerState -> withEffects(innerState, Effect.Share(innerState.character.urls.firstOrNull { it.type == "detail" }?.url)) }
    }

    override fun getInitialChange(): Change<InnerState, Effect?> {
        return asChange(InnerState(character, null, null, null, null, true, false))
    }

    override fun mapState(innerState: InnerState): State {
        val state = State(character = innerState.character,
                urls = innerState.character.urls.map {
                    ProcessedURLItem(it.type, { this::onUrlLinkClicked.invoke(it.url) });
                },
                comics = innerState.comics,
                series = innerState.series,
                stories = innerState.stories, events = innerState.events, loading = innerState.loading, error = innerState.error, callbackRunner = callbackRunner,
                clickListener = cardClickListener, shareListener = shareListener,
                showShare = innerState.character.urls.firstOrNull { it.type == "detail" } != null)
        Log.d("Vartika", "Character Detail mapState: $state")
        return state
    }

    class InnerState(var character: ProcessedMarvelCharacter, var comics: List<ProcessedMarvelComic>?, var series: List<ProcessedMarvelSeries>?, var stories: List<ProcessedMarvelStory>?, var events: List<ProcessedMarvelEvent>?, var loading: Boolean, var error: Boolean) {
        fun copy(): InnerState {
            return InnerState(character, comics, series, stories, events, loading, error)
        }
    }

    private fun loadComics(characterId: Int) {
        if (networkInterface == null) {
            return
        }
        val observable = networkInterface
                .loadCharacterComics(characterId)
                .onErrorReturn { throwable: Throwable? -> ArrayList() }
        val reducerObservable = observable
                .map<Reducer<InnerState, Effect>> { comics: List<ProcessedMarvelComic> ->
                    Log.d("Vartika3", "Reducer1: $comics")
                    Reducer { innerState ->
                        Log.d("Vartika3", "Reducer2: $comics")
                        val newInnerState = innerState.copy()
                        newInnerState.comics = comics
                        asChange(newInnerState)
                    }
                }
        enqueue(reducerObservable)
    }

    private fun loadSeries(characterId: Int) {
        if (networkInterface == null) {
            return
        }
        val observable = networkInterface
                .loadCharacterSeries(characterId)
                .onErrorReturn { throwable: Throwable? -> ArrayList() }
        val reducerObservable = observable
                .map<Reducer<InnerState, Effect>> { series: List<ProcessedMarvelSeries> ->
                    Log.d("Vartika3", "Reducer1: series $series")
                    Reducer { innerState ->
                        Log.d("Vartika3", "Reducer2: series $series")
                        val newInnerState = innerState.copy()
                        newInnerState.series = series
                        asChange(newInnerState)
                    }
                }
        enqueue(reducerObservable)
    }

    private fun loadStories(characterId: Int) {
        if (networkInterface == null) {
            return
        }
        val observable = networkInterface
                .loadCharacterStories(characterId)
                .onErrorReturn { throwable: Throwable? -> ArrayList() }
        val reducerObservable = observable
                .map<Reducer<InnerState, Effect>> { stories: List<ProcessedMarvelStory> ->
                    Log.d("Vartika3", "Reducer1: stories -> $stories")
                    Reducer { innerState ->
                        Log.d("Vartika3", "Reducer2: stories -> $stories")
                        val newInnerState = innerState.copy()
                        newInnerState.stories = stories
                        asChange(newInnerState)
                    }
                }
        enqueue(reducerObservable)
    }

    private fun loadEvents(characterId: Int) {
        if (networkInterface == null) {
            return
        }
        val observable = networkInterface
                .loadCharacterEvents(characterId)
                .onErrorReturn { throwable: Throwable? -> ArrayList() }
        val reducerObservable = observable
                .map<Reducer<InnerState, Effect>> { events: List<ProcessedMarvelEvent> ->
                    Log.d("Vartika3", "Reducer1: events -> $events")
                    Reducer { innerState ->
                        Log.d("Vartika3", "Reducer2: events -> $events")
                        val newInnerState = innerState.copy()
                        newInnerState.events = events
                        asChange(newInnerState)
                    }
                }
        enqueue(reducerObservable)
    }

    init {
        loadComics(character.id)
        loadSeries(character.id)
        loadStories(character.id)
        loadEvents(character.id)
    }
}